### DISCUSSION TITLE
Custom CellPlugins with inputs

### DISCUSSION THREAD

**Discussion by Max0nyshchenko** [react-page/react-page#1023] [Q&A]
Let's consider a situation when I want to have a plugin with input type checkboxes. So I'd expect in readOnly mode to be possible of modifying those checkboxes, checking them. But it's not possible. Is there a way to implement this behavior?

**Comment by macrozone** [comment]
@Max0nyshchenko what is the exact use case? I don't really understand your question.

**Comment by Max0nyshchenko** [comment]
I wrote custom plugin where added checkboxes input, and I want for other users to be able to check/modify them (checkboxes) in readOnly mode, but its not possible, content is not editable in readOnly mode.

Is it possible to make input fields editable (like textarea, input tags) but restict user from editing all other the content of the page?

**Comment by macrozone** [comment]
@Max0nyshchenko i still don't get your precise use case. 
Why would users in edit mode require to check checkboxes. What do they select exactly? Do these users modify the content or is it just some kind of interactive component?

You can use any component. So you can also mount a form or something like that. But i can't really help as its still unclear what you try to achieve

**Reply by Max0nyshchenko** [reply]
Issue now is that you cannot create forms, cause you will not be able to fill up them in the readOnly mode

**Reply by Max0nyshchenko** [reply]
onChange event is not implemented in HTMLRenderer

**Reply by Max0nyshchenko** [reply]
@macrozone

**Reply by macrozone** [reply]
hey @Max0nyshchenko ,

as mentioned, we need a clear use case to understand what you try to achieve.

It's best to first look at the two roles in a typical web-application (can be more in your case):

- you have an "Webmaster" that creates and edits content on your application. These users will use ReactPage in the edit mode
- you have "visitors" of the webapplication, these are the users of your customer,  they will just see content and maybe interactive components. ReactPage will display its content in read-only mode for these users

Then, consider what these "visitors" want to do on your webapplication. Probably get information and use a certain service.
Your "webmaster" now wants to have flexibility on what to display on the webapplication and he/she also wants to be able place some interactive components/widgets, let's say _for example a contact form_

Now let's assume that your "webmaster" wants to place a contact form on different pages, each time with a different set of inputs and with different recipients.

Its now your job as a developer to create a Component that takes the configuration that the "webmaster" wants as properties and that does the job that the "visitor" wants it to do (sending out an email). _This is something you have to do indipendently on whether you use ReactPage or any other solution_.

Let's assume you created such a Component, it might look like this:

````

const GenericContactForm = ({ fields, recipientId}) => {

  
return <Form onSubmit={formData => {
   sendContactemail(formData, recipientId) // this is your api call
}}>{
   fields.map(field => <FormField field={field} key={field.id} />
}</Form>


}
```

now you can use that component statically wherever you want like this:

```
<GenericContactForm recipientId="some-id" fields={["firstname", "lastname", "email"]} />


```

Now instead of adding that "hardcoded" into your application, you can add that as a `CellPlugin` to ReactPage:


```
const contactForm: CellPlugin<{fields: string[], recipientId: string}> = {
   id: "contactForm",
  version: 1,
  title: "A contact form",

  description: "A configurable contact form",
  Renderer: ({ data }) => {
    return (
      <GenericContactForm fields={data.fields} recipientId={data.recipientId} />
    );
  },
  controls: {
    type: "autoform",
    schema: {
      required: ["fields", "recipientId"],
      properties: {
        recipientId: {
          type: "string", 
          // its probably good to have a nice selection for the available recipients
          // i advice to not let webmasters enter an email-adress as this can be abused for spam
        },
        fields: {
          type: "array",
          items: {
            type: "string",
             // you can use an enum for all available fields
          },
        },
      },
    },
  },
}

```

------

this is an example of something that is interactive for visitors, but there is no need to have an onChange callback in readOnly-mode (and will never have as this is contrary to what a "readonly" mode is).

**Reply by macrozone** [reply]
@Max0nyshchenko i added an example of a (simple) form: https://react-page.github.io/examples/contactform

see the source code of the cellPlugin: https://github.com/react-page/react-page/blob/master/examples/plugins/contactForm.tsx

**Reply by Max0nyshchenko** [reply]
@macrozone Thank you for your great answer!

You got use-cases right. But what if I want that form data to be saved to Value JSON Schema and then reproduced again for other users?

**Reply by macrozone** [reply]
@Max0nyshchenko i don't know exactly, why you would want to do that (still waiting for a specific use case), but keep in mind that this would also mean that you would need to open up the api for both user groups to save the whole Value json. That means that even if you could limit the editor to just edit some part of the json, a malicious user could still edit the whole json through the api.
